pkg/cloud/services/compute/instance.go

```diff
@@ -584,16 +600,63 @@ func (s *Service) getImageIDFromName(imageName string) (string, error) {
 	}
 }
 
-// Helper function for getting image ID from name or ID.
-func (s *Service) getImageID(imageUUID, imageName string) (string, error) {
-	if imageUUID != "" {
-		// we return imageUUID without check
-		return imageUUID, nil
-	} else if imageName != "" {
-		return s.getImageIDFromName(imageName)
+func (s *Service) GetServerGroupID(serverGroupID string, serverGroupFilter *infrav1.ServerGroupFilter) (string, error) {
+	if serverGroupFilter == nil {
+		if serverGroupID == "" {
+			return "", nil
+		}
+
+		// only fallback to the legacy value if a filter wasn't provided
+		serverGroupFilter = &infrav1.ServerGroupFilter{
+			ID: serverGroupID,
+		}
 	}
 
-	return "", nil
+	// if we have an ID, use that
+	if serverGroupFilter.ID != "" {
+		serverGroup, err := s.getComputeClient().GetServerGroup(serverGroupFilter.ID)
+		if err != nil {
+			return "", err
+		}
+		if serverGroupFilter.Name != "" && serverGroupFilter.Name != serverGroup.Name {
```

Comment:
should we really have this check? or just safely ignore the Name ? I think the logic here want to make sure they are consistent..

Reply:
So while I think this is silly check and I think providing both a name and ID is even sillier, but if someone does it then I think we should respect their request fully. To not do so would be confusing.

Reply:
I'm with @jichenjc on this one. Can we not just add an admission validation that if you specify ID you can't specify anything else?

If the machine creation fails because the specified ID was invalid that should be reported in the machine status.